Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -183,58 +184,113 @@ public override OrderEvent MarketFill(Security asset, MarketOrder order)
+        /// In this case, a new bar opens below/above the stop trigger price, and the order is filled with the opening price.
         /// <seealso cref="MarketFill(Security, MarketOrder)"/>
         public override OrderEvent StopMarketFill(Security asset, StopMarketOrder order)
         {
-            //Default order event to return.
+            // Default order event to return.
             var utcTime = asset.LocalTime.ConvertToUtc(asset.Exchange.TimeZone);
             var fill = new OrderEvent(order, utcTime, OrderFee.Zero);
 
-            //If its cancelled don't need anymore checks:
+            // If cancelled, don't need anymore checks:
             if (order.Status == OrderStatus.Canceled) return fill;
 
-            // make sure the exchange is open/normal market hours before filling
+            // Make sure the exchange is open/normal market hours before filling
             if (!IsExchangeOpen(asset, false)) return fill;
 
-            //Get the range of prices in the last bar:
-            var prices = GetPricesCheckingPythonWrapper(asset, order.Direction);
-            var pricesEndTime = prices.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+            // Get the range of prices in the last bar:
+            var tradeOpen = 0m;
+            var tradeHigh = 0m;
+            var tradeLow = 0m;
+            var endTimeUtc = DateTime.MinValue;
 
-            // do not fill on stale data
-            if (pricesEndTime <= order.Time) return fill;
+            var subscribedTypes = GetSubscribedTypes(asset);
 
-            //Calculate the model slippage: e.g. 0.01c
-            var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                var trades = asset.Cache.GetAll<Tick>().Where(x => x.TickType == TickType.Trade && x.Price > 0);
 
-            //Check if the Stop Order was filled: opposite to a limit order
+                foreach (var trade in trades)
+                {
+                    tradeOpen = tradeOpen == 0 ? trade.Price : tradeOpen;
+                    tradeHigh = Math.Max(tradeHigh, trade.Price);
+                    tradeLow = tradeLow == 0 ? trade.Price : Math.Min(tradeLow, trade.Price);
+                    endTimeUtc = trade.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+
+                if (tradeBar != null)
+                {
+                    tradeOpen = tradeBar.Open;
+                    tradeHigh = tradeBar.High;
+                    tradeLow = tradeBar.Low;
+                    endTimeUtc = tradeBar.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+                }
+            }
+
+            // Do not fill on stale data
+            if (endTimeUtc <= order.Time) return fill;
```

Comment:
I think we can eventually move all this logic above to a shared method given it's being reused, can be in it's own PR with no behavior change 👍